packages/subproviders/src/subproviders/nonce_tracker.ts

```diff
@@ -0,0 +1,107 @@
+import * as _ from 'lodash';
+
+import EthereumTx = require('ethereumjs-tx');
+import ethUtil = require('ethereumjs-util');
+import providerEngineUtils = require('web3-provider-engine/util/rpc-cache-utils');
+
+import { JSONRPCPayload } from '../types';
+
+import { Subprovider } from './subprovider';
+
+const NONCE_TOO_LOW_ERROR_MESSAGE = 'Transaction nonce is too low';
+
+export class NonceTrackerSubprovider extends Subprovider {
+    private _nonceCache: { [address: string]: string } = {};
+    private static _reconstructTransaction(payload: JSONRPCPayload): EthereumTx {
+        const raw = payload.params[0];
+        const transactionData = ethUtil.stripHexPrefix(raw);
+        const rawData = new Buffer(transactionData, 'hex');
+        return new EthereumTx(rawData);
+    }
+    private static _determineAddress(payload: JSONRPCPayload): string {
+        switch (payload.method) {
+            case 'eth_getTransactionCount':
+                return payload.params[0].toLowerCase();
+            case 'eth_sendRawTransaction':
+                const transaction = NonceTrackerSubprovider._reconstructTransaction(payload);
+                return `0x${transaction.getSenderAddress().toString('hex')}`.toLowerCase();
+            default:
+                throw new Error('Invalid Method');
+        }
+    }
+    constructor() {
+        super();
+    }
+    // tslint:disable-next-line:async-suffix
+    public async handleRequest(
+        payload: JSONRPCPayload,
+        next: (callback?: (err: Error | null, result: any, cb: any) => void) => void,
```

Comment:
we can pull out some of these function signatures as type declarations at the top of the file for readability